File: Bootcamp/week-3/121-BestTimeToBuyAndSellStock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minimum, profit = prices[0], 0
        for i in range(len(prices)):
            profit = max(profit, prices[i] - minimum)
            if prices[i] < minimum:
                minimum = prices[i]

        return profit


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profits = {}
        max_profit = 0

        def profit(day):
            nonlocal max_profit
            if day == len(prices) - 1:
                return prices[day]

            if profits.get(day):
                return profits[day]
            next_max = max(prices[day], profit(day + 1))
            max_profit = max(max_profit, next_max - prices[day])
            profits[day] = next_max

            return profits[day]

        profit(0)
        return max_profit

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.c
#define MAX(a, b) (((a) > (b)) ? (a) : (b))

int maxProfit(int* prices, int pricesSize){
    int profit = 0, bestBuy = INT_MIN;   // bestBuy keeps the best day we spend minimum if we buy
    
    bestBuy = profit - prices[0]; //if we buy on the first day we'll have a negative profit
    
    for(int i = 1; i < pricesSize; i++){
        // on the subsquent days we'll have three choices 
        // 1 - to determine if it's the best day to buy rather than the previous best day i.e. 0 - prices[i]
        // 2 - to check if the profit is better if we sell i.e. bestBuy + prices[i] ----> bestbuy is cost and prices[i] is gain
        // 3 - to do nothing 
        
        profit = MAX(profit, prices[i] + bestBuy); // choice 2 if profit gets bigger else choice 3 
        bestBuy = MAX(bestBuy, -prices[i]); // choice 1
    }
    return profit;
}